### DISCUSSION TITLE
Rendering midi files via server daemon

### DISCUSSION THREAD

**Discussion by johnkolen** [FluidSynth/fluidsynth#779] [Q&A]
This is probably a silly question. I'm currently rendering midi files via invocations of the fluidsynth command line (e.g. fluidsynth -f script tune.midi). I'd like to switch over to using server mode. While I'm able to send all my settings across port 9800, I don't know how to trigger rendering the midi file (fluid_player_add/fluid_player_play) over the TCP connection. Is there a fluidsynth shell command or setting that will let me specify the file and start processing it?

**Comment by mawe42** [comment]
Controlling the player via the fluid shell is something @jjceresa has implemented recently here: https://github.com/FluidSynth/fluidsynth/pull/713

You could try out the current master to see if that fits your needs.

Cheers
Marcus

**Reply by johnkolen** [reply]
Thanks for the pointer. While that PR gives you control over starting and stoping rendering, it's only part of what I was looking for. My intention was to run the fluidsynth server, connect to it and provide a midi file so it could render a wave file. It appears that the only way to specify the midi file is on the command line. Likewise, the command line is the only way to specify the output file as well (Warning: 'audio.file.name' is not a realtime setting, changes won't take effect.)

It would be nice to be able evaluate a script like this:
set audio.driver file
set audio.file.type wav
set audio.file.name output.wav
load SoundFonts/FluidR3_GM.sf2
player_add input.midi
player_start

**Reply by derselbst** [reply]
Since you want to use that instance for file rendering only, why do you have to write commands like `set audio.driver` in a command file? Just start it with the right flags from command line in the first place.
Also, the `audio.file.name` cannot be changed afterwards, so why putting this in a command file as well?
Your idea of starting and stopping the rendering doesn't work that way: The audio driver will always render audio. If the player is stopped, it will render silence.
And why are you using the `file` driver? The time to render will be as long as the play-duration of the MIDI.

> Likewise, the command line is the only way to specify the output file as well

Exactly.

**Reply by johnkolen** [reply]
My use case involves converting lots of midi files to wave on demand. Instantiating a new instance of fluidsynth incurs a latency penalty as well as creates a memory bottleneck. It would be nice to have a single instance running as a server to catch rendering requests. From what you are saying, it looks like I'll have to roll my own. Also, I should point out that rendering takes about 15% of the play duration with the file driver.

**Reply by mawe42** [reply]
If it's mainly the Soundfont loading that is causing latency and memory problems with the simple fluidsynth binary, you could try the dynamic sample loading option. It drastically reduces initial startup latency and memory footprint by only loading the presets that are actually needed to render the file. But if that still doesn't fit your use-case, using the FluidSynth API to write your own server is probably the way to go.
